analyses/batch-effects/util/functions.R

```diff
@@ -0,0 +1,361 @@
+  
+  # filter out NAs
+  data_t = filter(data_t, !is.na(data_t$batch))
+  
+  return(data_t)
+}
+
+# for batchQC to work, the batches must be numeric values
+batch_prep = function(data_t){
+  # verify that batch is indeed a factor and numeric
+  batch = as_factor(data_t$batch)
+  batch = as.numeric(batch)
+  
+  # make so that batch values start at 1
+  levels(batch) <- 1:length(levels(batch))
+  return(batch)
+}
+
+covariate_filtering = function(covariate, batch_column){
+  
+  # Only grab RNA-Seq data
+  covariate = filter(covariate, experimental_strategy == "RNA-Seq")
+  
+  # Add a new column called batch which will be a duplicate of the whichever column name is passed into the function
+  covariate$batch = unlist(covariate[which(names(covariate) == batch_column)])
+  
+  # convert 'batch' column into numbers through first converting it into a factor
+  covariate$batch = as_factor(covariate$batch)
+  
+  # move the batch column to the 4th location (optional step which is useful for debugging)
+  covariate = covariate %>% select(1:4, batch, everything())
+  
+  return(covariate)
+}
+
+# read in the covariate data and create a tibble with two columns: the patient ID and the Batch
+make_id_batch = function(covariate, batch_column){
+  
+  # move the batch column to the front and only include RNA-Seq data
+  covariate = covariate_filtering(covariate, batch_column) 
+
+  # make a new matrix which will contain patient ID's and the corresponding batch
+  id_batch = cbind("Kids_First_Biospecimen_ID" = covariate$Kids_First_Biospecimen_ID, "batch" = covariate$batch)
+
+  # rename columns
+  colnames(id_batch) <- c("Kids_First_Biospecimen_ID", "batch")
+
+  # convert id_batch from a matrix to a tibble
+  id_batch = as_tibble(id_batch)
+  
+  return(id_batch)
+}
+
+make_id_batch_histology = function(covariate, batch_column){
+  # move the batch column to the front and only include RNA-Seq data
+  covariate = covariate_filtering(covariate, batch_column) 
+  
+  # make a new matrix which will contain patient ID's and the corresponding batch and histology
+  id_batch_histology = cbind("Kids_First_Biospecimen_ID" = covariate$Kids_First_Biospecimen_ID, "batch" = covariate$batch, "histology" = covariate$short_histology)
+  
+  #rename columns
+  colnames(id_batch_histology) <- c("Kids_First_Biospecimen_ID", "batch", "histology")
+  id_batch_histology = as_tibble(id_batch_histology)
+  
+  # convert id_batch_histology from a matrix to a tibble
+  id_batch_histology = as_tibble(id_batch_histology)
+  
+  return(id_batch_histology)
+}
+
+# make two PCA plots before and after applying combat
+make_before_and_after_combat_plots = function(data_t, batch, report_name, run_combat) {
+  
+  data_t$batch = batch_prep(data_t)
+  
+  #make a matrix representing just the gene expression data (this gets plugged into combat)
+  poly1matrix = t(data_t[,4:ncol(data_t)])
+  
+  pca = prcomp(log2(data_t[,4:ncol(data_t)]+ 1))
```

Comment:
It looks like you have to do this transformation twice: once before ComBat and once for this call to `prcomp()`. 

Instead, what if you did your `log2()` step elsewhere (e.g., in one of your cleaning functions), before `data_t` gets to this function?

I'd also consider whether or not you need to transpose before you get to the `prcomp()` steps.